stream_alert/rule_processor/rules_engine.py

```diff
@@ -154,6 +157,81 @@ def match_event(cls, record, rule):
         return True
 
     @classmethod
+    def match_types(cls, record, normalized_types, datatypes):
+        """Match normalized types against record
+
+        Args:
+            record (dict): Parsed payload of any log
+            normalized_types (dict): Normalized types
+            datatypes (list): defined in rule options, normalized_types users
+                interested in.
+
+        Returns:
+            (dict): A dict of normalized_types with original key names
+
+        Example 1:
+            datatypes=['defined_type1', 'defined_type2', 'not_defined_type']
+            This method will return an empty dictionary and log datatypes
+                "not defined" error to Logger.
+
+        Example 2:
+            datatypes=['defined_type1', 'defined_type2']
+            This method will return an dictionary :
+                {
+                    "defined_type1": [original_key1],
+                    "defined_type2": [[original_key2, sub_key2], original_key3]
+                }
+        """
+        results = dict()
+        if not (datatypes and cls.validate_datatypes(normalized_types, datatypes)):
+            return results
+
+        for key, val in record.iteritems(): # pylint: disable=too-many-nested-blocks
```

Comment:
hah, that pylint warning exists for a reason :)

Here's how to avoid having too much nesting: https://blog.rburchell.com/2010/11/coding-antipatterns-excessive-nesting.html